### ISSUE TITLE
Grep changes to wrong buffer when quitting denite.nvim

### ISSUE THREAD

**Issue by wilywampa** [Shougo/denite.nvim#344]
# Problems summary

Grep result switches buffer on exit

## Expected

Should stay at the grep result after quitting denite.nvim

## Environment Information (Required!)

 * denite version(SHA1): f112b16

 * OS: macOS Sierra

 * Vim/neovim version: NVIM v0.2.1-596-g5a0acd09c Build type: Release

 * `:CheckHealth` result(neovim only):

```
health#nvim#check
========================================================================
## Configuration
  - SUCCESS: no issues found

## Performance
  - SUCCESS: Build type: Release

## Remote Plugins
  - SUCCESS: Up to date

## terminal
  - INFO: key_backspace (kbs) terminfo entry: key_backspace=^H
  - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~

## tmux
  - SUCCESS: escape-time: 50ms
  - INFO: $TERM: screen-256color

health#provider#check
========================================================================
## Clipboard (optional)
  - SUCCESS: Clipboard tool found: pbcopy

## Python 2 provider (optional)
  - INFO: `g:python_host_prog` is not set.  Searching for python2 in the environment.
  - SUCCESS: pyenv found: "/usr/local/Cellar/pyenv/1.0.7/libexec/pyenv"
  - INFO: Executable: /usr/local/bin/python2
  - INFO: Python2 version: 2.7.13
  - INFO: python2-neovim version: 0.1.13
  - SUCCESS: Latest python2-neovim is installed: 0.1.13

## Python 3 provider (optional)
  - INFO: `g:python3_host_prog` is not set.  Searching for python3 in the environment.
  - SUCCESS: pyenv found: "/usr/local/Cellar/pyenv/1.0.7/libexec/pyenv"
  - INFO: Executable: /Users/jniehus/Documents/dev/py3.6.1env/bin/python3
  - INFO: Python3 version: 3.6.1
  - INFO: python3-neovim version: 0.1.13
  - SUCCESS: Latest python3-neovim is installed: 0.1.13

## Ruby provider (optional)
  - INFO: Ruby: ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin16]
  - INFO: Host: /usr/local/bin/neovim-ruby-host
  - WARNING: Gem "neovim" is out-of-date. Installed: 0.4.0, latest: 0.5.0
    - SUGGESTIONS:
      - Run in shell: gem update neovim
```

## Provide a minimal init.vim with less than 50 lines (Required!)

```vim
scriptencoding utf8
set runtimepath+=/Users/jniehus/Downloads/dev/denite.nvim
nnoremap <C-g> :<C-u>Denite grep -no-quit<CR>
```

## The reproduce ways from neovim starting (Required!)

 1. Start nvim from the denite.nvim directory with `nvim README.md`
 2. Start `:Denite grep` by pressing `<C-g>` and search for "git" as the pattern
 3. Select one of the results in any file except `README.md`
 4. Go to normal mode and exit denite.nvim with `<C-o>q`
 5. The current buffer will be `README.md` instead of the selected candidate

**Comment by Shougo** [comment]
This is intended behavior.
Because, ....
https://github.com/Shougo/denite.nvim/issues/324#issuecomment-316123547

> Cursor should be where it was before exiting denite.nvim, e.g. at the most recent line that was jumped to from grep

You have closed the denite buffer.
The cursor should be where it was before exiting denite.nvim.

**Comment by Shougo** [comment]
> Should stay at the grep result after quitting denite.nvim

Conflicts!

**Comment by wilywampa** [comment]
OK. Let's say I am in some random buffer I don't care about and I want to search for a line of code in  a different file. I run `:Denite grep` and jump to that line of code. I am where I want to be and all is right in the world. Then I close the denite.nvim window and render the whole exercise pointless by going back to the random buffer I don't care about? That's the intended behavior?

**Comment by Shougo** [comment]
I can change the behavior.
But you cannot reset the buffer when https://github.com/Shougo/denite.nvim/issues/324.
If you want fix #324, I cannot fix #344.

This is you have selected the behavior!

**Comment by Shougo** [comment]
#324 or #344 please choice the behavior.
You cannot choose both.

**Comment by wilywampa** [comment]
It seems to me that #344 is a duplicate of #324 and that #324 was fixed but broke again in the past few days. I don't understand how #324 conflicts with #344. Both want the active buffer with the search result to remain current after closing denite.nvim.

**Comment by Shougo** [comment]
I don't want to explain it though....
I have not much free time.
:-(

Please see #344 more correctly.

Note: `<denite:quit>` moves to the previous buffer.
To fix #344 problem, denite must overwrite the previous buffer to the selected candidate.

> Select any result and jump to it - the cursor should be back in the denite window now

You have selected the candidate.  So the previous buffer is the selected candidate.
It is your desired behavior in #344.

> Quit the denite buffer by pressing <C-c> - this will reset the cursor position in whatever file had the grep result

So if you have quit the buffer, it will be the candidate buffer instead of the previous buffer. 

Do you understand?

**Comment by Shougo** [comment]
I should change `-no-quit` behavior.

**Comment by Shougo** [comment]
> It seems to me that #344 is a duplicate of #324 and that #324 was fixed but broke again in the past few days.

It is not.

**Comment by wilywampa** [comment]
>Note: `<denite:quit>` moves to the previous buffer.
To fix #344 problem, denite must overwrite the previous buffer to the selected candidate.

Doing that is what makes sense to me. Jumping to a candidate in grep results should set that candidate as the place the cursor will go after `<denite:quit>`. Sorry for all the confusion!

**Comment by Shougo** [comment]
OK.  I will do.
